PR #2: eventData --> master

harishbommakanti opened the pull request:
added features for adding events for specific seasons, and teams for the specific events

harishbommakanti committed:
Initial Commit for EventData

- added event data js/edited the html file
- added a simple text at the top indicating which season the user is entering events for, if any are saved

harishbommakanti committed:
Added the js/html to add events

- code basically mirrors the addSeason function in the seasonConfigScript.js
- user is able to enter events for the current (last checked) season on previous page
- now, work needs to be done on entering elements/code for adding/displaying teams for each event

harishbommakanti committed:
starter code for event specific teams

- added base html features, mirrors the teamElements on the seasnConfig page

harishbommakanti committed:
most of team enter code done

- most of code mirrors the event code from the season config page, which is why this is a huge commit
- focused on adding helper methods
- edge case in season config fixed
- need to work on making the cookies not overwriting and making the table update onload and onchange of the event

harishbommakanti committed:
Overriding team names/numbers code

- made it so that the user can override a currently existing team with a number, or vice versa. Looking to add this feature to the seasonConfig elements as well
- user can't enter same team twice
- commenting
- next steps are working on deleting specific teams and deleting all teams

harishbommakanti committed:
deleteAllTeams code

- short method for deleting all teams of current event

harishbommakanti committed:
Delete Specific Table items feature added

- user can now delete specific teams in event data page

harishbommakanti committed:
delete specific element feature added to season config

- on season config page, user can now delete a specific element only
- took away the deleteAllElements features for both pages as a misclick may destroy all the data

harishbommakanti committed:
commenting/code organization

- added commenting here and there and organized some code

harishbommakanti committed:
minor code fix

- when user deletes event, the cookie with the teams under that event are also deleted

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,388 @@
+//credit to w3schools for js tutorials, stackoverflow
+
+function loadPage()
```
why do you have both a onload="loadPage" and a $(window).ready ?

Try to only have one. (in this case, probably remove `onlocad="loadPage"` from the html and call `loadPage()` at the start of the .ready function (and/or move more of that code block into this function).

harishbommakanti replied:
I'll call the loadpage method from the $(window).ready block

Also for the second comment about overwriting the team number but not the name, I had that as a feature in case someone messed up and didn't want to go through the process of deleting the team and then adding it. I basically said if the team exists but the number is different, override the number but keep the name and vice versa, but if nothing matches, then add it as a new team. Conversely, if everything matches, then say 'already stored'. I'll look more into it though.

I do need to revise the deleteTeam (eventData) and deleteElement (seasonConfig) methods, it seems that when I have nothing in the box, everything just disappears, like you were mentioning.

mshafer1 replied:
You should make overwriting only happen on exact team number match. (like, it's a good idea, but there's a lot of team numbers that are similar, and I think that's what confused me).

harishbommakanti replied:
Ok, makes sense. It'll only overwrite the team number if the team number input matches. And if the name already exists, it'll say 'team x already exists'

mshafer1 replied:
also, don't error on duplicate team names for different numbers - it is possible. Only numbers are guaranteed to be unique.

mshafer1 replied:
ok, I think you have a mix up on editing teams:

![image](https://user-images.githubusercontent.com/2565361/68258240-bfb9a680-fffb-11e9-9054-975edee3255d.png)

I got into this state by trying to change team 123 from named "test" to "test1", this yields showing only "test1" in the list (dropping test2), and the change up of team numbers vs names in the cookies makes me think that you have an issue of assignment location.

mshafer1 replied:
You should maybe consider using a dictionary, 

```js
Teams = {
123: "test1",
456: "test2",
}

set_cookie('/event_data teams/' + currCheckedSeason + '/teams', JSON.stringify(Teams));

raw_stored_data = get_cookie(('/event_data teams/' + currCheckedSeason + '/teams');
teams = JSON.parse(raw_stored_data);
teams[123] // test1
```

then, updating OR adding a team is 
```js
Teams[team_number] = team_name;
set_cookie('/event_data teams/' + currCheckedSeason + '/teams', JSON.stringify(Teams));
```

harishbommakanti replied:
Thank you! I didn't think of that but the dictionary works perfectly in this case as it doesn't allow for duplicate team numbers or keys.

harishbommakanti committed:
Adding a continue to next page button, minor fixes

- edited info on front page
- edited text on a button on season config
- entered a 'continue to team data' button to event data

mshafer1 reviewed (changes requested):
I had some issues running this so I'm not looking very far into the code just yet.

I tried to add a team then another team then delete the first and the entire poofed (emptied), then I tried re-adding my missing team and got "team already added".

I think you have some logic errors around adding and removing teams to and event.
You might want to look into Jest as a unit testing framework (it doesn't handle the cookie and DOM content, so writing JS code to make it testable requires moving the logic pieces into their own methods that you test and combine for the cookie and DOM manipulation.

mshafer1 commented:
It also looks like if I load the page and insert a team, that it overwrites the team number of the already stored team and drops (doesn't do anything with) the team name.

harishbommakanti committed:
fixed deleteSpecificTeam/Element

- fixed case where input is '', shouldn't make all elements/teams dissapear

harishbommakanti committed:
moved loadPage to window.ready block

- removed redundant onLoad=loadPage

harishbommakanti committed:
Changed team overwriting methods

- now, only if the team number matches, the associated name for that team will change to the new name
- if a name already exists, the program will not overwrite the number to the new number, but will just say 'name x already exists'
- minor changes to the text on season config page and event data page

mshafer1 reviewed (commented):
It's unclear that the selected event is associated with the table, I was able to "guess" it based on the season config page, but you should make it more explicit (maybe, "select Event to add teams for", or similar)

harishbommakanti committed:
curr event shown near table

- now user won't have to 'guess' the event for which the teams are being added to

harishbommakanti committed:
used a JSON object to store event teams

- using a JSON object simplifies a lot of the code
- adding teams/updating the table should be ok, will further test
- need to modify the deleting teams feature

harishbommakanti committed:
added delete method

- delete method works now
- deleted old code (can just revert if this method doesn't work)
- added background as an image instead of referencing a url

harishbommakanti committed:
minor changes

- some text changes on the html pages

harishbommakanti commented:
> It's unclear that the selected event is associated with the table, I was able to "guess" it based on the season config page, but you should make it more explicit (maybe, "select Event to add teams for", or similar)

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    if (isSeasonSaved())
+    {
+        var currSeason = getCookie('currCheckedSeason');
+        finalText = 'Here, enter all the events that the team will attend for season "' + currSeason + '".';
```
or will be scouting from - we can scout using the live stream.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+        finalText = 'Here, enter all the events that the team will attend for season "' + currSeason + '".';
+    } else
+    {
+        finalText = 'No season is currently checked. Go back to the Season Configuration page ' + 
```
if we should have to select a season, provide a link to the season page and leave the rest of this page disabled - maybe make this a modal message (see w3css's w3-modal for a good way to do that) or put in loadPage() to disable all inputs and in this method enable them in the true case.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+            for (var i = 0; i < eventList.length; i++)
+            {
+                var cookieVal = eventList[i];
+                if (!$('#eventRadioList').html().includes(cookieVal))
```
hmmm. . .  this could be problematic if two events have a subset name (where one's name is contained entirely in another's name), . . . but I'm not sure that happens in real life . . .

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+//event cookies --> /eventData/{eventNames}/Austin District EventΔPasadena...
+function addNewEvent()
+{
+    var newEvent = $("#newEventInputBox").val().trim().toString();
```
it should already be a sting :)

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    var newEvent = $("#newEventInputBox").val().trim().toString();
+    var currSeason = getCookie('currCheckedSeason');
+
+    if (currSeason === " ") //if season is blank don't continue with function
```
`currSeason.trim().length == 0 `
will be true if currSeason is " " or "", so it might work better (to handle the case of cookie not set).

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    if (isAlreadyStored) //if season is already stored
```
```js
if(isAlreadyStored)
{
$("#newEventConfirmation").html("<i>Event \'" + newEvent +"\' already exists.</i>"); 
return;
}

//add to radioList
$("#newEventConfirmation").html("<i>Event " + "\'"+ newEvent +"\'"+" has been added.</i>"); 
```

exiting the function after handling rather than using else blocks cuts down on the nesting - which equals easier to follow the expected case, but this is a stylistic opinion, code works the same either way, so you decide.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    var currCheckedEvent = getCurrEvent();
+    var cnameEnding = currCheckedSeason + '/' + currCheckedEvent;
+    
+    //splices the eventList cookie and gets rid of the function to be deleted
```
function to be deleted? do you mean event?

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    $("#eventSpecificTeamTable").html('<tr><th><b>Team Number</b></th> <th><b>Team Name</b></th> </tr>');
+
+    var teamsJSON = JSON.parse(getCookie(event_data_cookie_name(season,currEvent)));
+    if (teamsJSON == '{}') return;
```
after calling `JSON.parse`,
it's not JSON anymore, just an object (or dictionary)

```js
var teams = JSON.parse...
if(Object.keys(teams).length == 0) {
return;
}
```

`Object.keys` is a prototype method that can be called  on any method, returns an array.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    for (var i = 0; i < pairs.length; i++)
+    {
+        var teamNumber = pairs[i][0];
+        var teamName = pairs[i][1];
```
this does work, but I wonder if this would be simpler or easier to read:

```js
var keys = Object.keys(teamsJSON); // see above on variable name
for(var i = 0; i < keys; i++) { 
  var teamNumber = keys[i];
  var teamName = teamsJSON[teamNumber];
}
```

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    var season = getCookie('currCheckedSeason');
+    var currEvent = getCurrEvent(); 
+    var cname = event_data_cookie_name(season,currEvent);
+    var teamsJSON = JSON.parse(getCookie(cname));
```
you're doing this
```js
var teamsJSON = JSON.parse(getCookie(cname));
```

a lot - make it a function:

```js
function getTeams(season, currevent) {
var cname = event_data_cookie_name(season,currEvent);
var result = JSON.parse(getCookie(cname));
return result;
}
```

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    delete teamsJSON[input];
+    setCookie(cname,JSON.stringify(teamsJSON),750);
+    setTeamTable(season,currEvent);
+    $("#deleteSpecificTeamTextBox").val("").focus();
```
good! - makes it easy to add multiple teams quickly

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+
+    //event specific teams code
+    $('#teamName').keypress(function(e){
+        if (e.keyCode == Enter_key_code) $('#teamSaveButton').click();
```
call the method the click calls rather than simulating a click.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+        var teamName = $("#teamName").val().trim();
+        var teams = {};
+
+        if (teamNumber === "" || teamName === "")
```
this is a pretty long function, go ahead and put it outside loadPage and use the name here:

```js
function saveTeam() {
...
}
...
   $('#teamSaveButton').click(saveTeam);
...
```

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+            return;
+        }
+
+        if (currSeason === ' ')
```
what happens if the cookie doesn't exist?

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+        } else //cookie not stored yet
+        {
+            teams = {};
+        }
```
just set the default up at the start:

```js
var raw_stored_data = getCookie(cname).tim();
var teams = {};
if (raw_stored_data.length > 0) //if cookie is stored yet
{
    teams= JSON.parse(raw_stored_data);
}
```

Also, `(!raw_stored_data == '')` probably isn't doing what you think.
`!raw_stored_data` will convert `raw_stored_data` to a bool then not it (a string other than empty = true, and empty string = false), ...
that's functionally equivalent to, `(raw_stored_data == "")?false:true`
then, it will convert the `''` at the end to a bool (which will be false) for the comparison, so what you've coded is:
`if( (raw_stored_data == "")?false:true == false)`, this is doing what you wanted, but it would read a lot better with either:
`if( raw_stored_data != '')` or `if(raw_stored_data.length > 0)`

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    var re = new RegExp(name + "=([^;]+)");
+    var value = re.exec(document.cookie);
+    return (value != null) ? unescape(value[1]) : " ";
+}
```
put these three methods in a common js file (maybe call it cookieHelpers.js) and include it from the html rather than copy and pasting it to multiple pages.

mshafer1 commented on eventDataScript.js:
```diff
@@ -0,0 +1,296 @@
+    return (value != null) ? unescape(value[1]) : " ";
+}
+
+function event_data_cookie_name(season,event)
```
are these copy-pasted? if so, move them (including the original copies) to a common js file (maybe call it eventHelpers.js) and include it from the html. That way, if you wind up changing how something works (like the path), you only have to change one copy.

mshafer1 commented on season-config.html:
```diff
@@ -47,7 +47,7 @@ <h3>Enter in any scoring opportunities in the first text box
 
         <fieldset id = "scoringOpportunities">
```
Just a thought, how would we track notes or other free text fields?, probably keep this as is, but put on the "future improvements list" to allow for selecting the input type.

mshafer1 commented on seasonConfigScript.js:
```diff
@@ -20,7 +20,7 @@ function addNewSeason()
     if (document.cookie.length > 0) //if cookies are stored at all
     {
-        if (getCookie("seasonList") !== "")
+        if (getCookie("seasonList") !== " ")
```
`getCookie("seasonList").strip().length > 0`
would be more consistent, doesn't care about whitespace or not.

mshafer1 commented on seasonConfigScript.js:
```diff
@@ -246,7 +256,7 @@ $(document).ready(function()
         {
             var season = getCookie('currCheckedSeason');//$("input[name=seasonListItem]:checked","#seasonRadioList").val();
-            if (typeof(season) !== "undefined")
+            if (season !== " ")
```
`season.strip().length > 0` is the more standard way of doing this.

mshafer1 commented on styles.css:
```diff
@@ -1,7 +1,7 @@
 #background {
     height: 100%;
-    background-image: url(https://utemplates.net/wp-content/uploads/2017/07/White-Textured-Background.jpg);    ;
+    background-image: url('images/background.jpg');
```
👍

mshafer1 reviewed (approved):
seems to work pretty well, made some comments on the code for how you can have better (more maintainable, or easier to read) code.

harishbommakanti committed:
changes from PR#2 to make code readable/maintanable

- all changes that Mr. Matt suggested
- last item to do is work with modal messages and disable the rest of the event data page if the season is not selected

harishbommakanti committed:
modal box added

- if no season, then a modal box will appear and all the inputs on the events data page will be disabled
- inputs are not disabled once season is checked on previous page

harishbommakanti committed:
fixed minor glitches in seasonConfig

- minor edge cases and stuff I didn't catch before, all event data stuff should be good

mshafer1 commented on event-data.html:
```diff
@@ -25,12 +26,22 @@ <h1 id = "title">FRC 6321 Scouting</h1>
+        <div id= 'seasonAlertModal' class='modal'>
+            <div class='modal-content'>
+                <span class='close' onclick = 'closeSeasonModal()'>&times;</span>
+                <p>No season is currently checked. Go to the <a href="season-config.html">Season Configuration</a> page 
```
```js
<p>No season is currently checked. Go to the <a href="season-config.html" class="r3-button">Season Configuration</a> page 
```

keep the appearance of buttons/links consistent.

mshafer1 commented on scripts/eventDataScript.js:
```diff
@@ -17,20 +17,25 @@ function isSeasonSaved()
-        'to enter new seasons and season-specific elements/pit scouting questions. Make sure you check the new season you enter.'
+        var finalText = 'Here, enter all the events that the team will be scouting from for season "' + currSeason + '".';
+        $("#eventDataIntroText").text(finalText);
+        $('.disableIfNoSeason').prop("disabled", false) //remove disable on inputs
```
if you change the disable below, update this as well.

mshafer1 commented on scripts/eventDataScript.js:
```diff
@@ -17,20 +17,25 @@ function isSeasonSaved()
-    $("#eventDataIntroText").text(finalText);
+
+    $('#seasonAlertModal').show(); //show modal
+    $('.disableIfNoSeason').prop("disabled", true); //disable all inputs on page
```
`$(':input').prop('disabled', true)` would get the buttons too. (just a suggestion, try it first).

mshafer1 commented on scripts/eventDataScript.js:
```diff
@@ -73,50 +78,46 @@ function changeCurrEvent()
     {
-        if (cookieName !== "")
+        var val = eventNames[i].trim();
+        if (newEvent==val.toString())
```
it's already a string :)

mshafer1 commented on scripts/eventDataScript.js:
```diff
@@ -192,98 +194,74 @@ function deleteSpecificTeam()
+    var teamName = $("#teamName").val().trim();
+    var teams = {};
+
+    if (teamNumber === "" || teamName === "")
```
`teamNumber` will never triple equal empty string, triple equals checks type as well.

if #teamNumber.val() is empty,
`var teamNumber = parseInt($("#teamNumber").val().trim());`
will yield `NaN` (not a number), to check for this, use `Number.isNaN()`:
```js
if(Number.isNaN(teamNumber) || teamName === "")
```

mshafer1 commented on scripts/eventDataScript.js:
```diff
@@ -192,98 +194,74 @@ function deleteSpecificTeam()
+    {
+        $("#teamConfirmationBox").html("<i>Please check a season.<\i>");
+        return;
+    } else if (currEvent === ' ')
```
`currEvent.trim().length == 0`

mshafer1 commented on styles.css:
```diff
@@ -138,4 +138,44 @@ table, th, td
 {
   width: 100%;
   resize:none;
+}
+
+/*for the season alert on the event data page*/
+/* The Modal (background) */
+.modal {
+  display: none; /* Hidden by default */
+  position: fixed; /* Stay in place */
+  z-index: 1; /* Sit on top */
+  left: 0;
+  top: 0;
+  width: 100%; /* Full width */
+  height: 100%; /* Full height */
+  overflow: auto; /* Enable scroll if needed */
+  background-color: rgb(0,0,0); /* Fallback color */
+  background-color: rgba(0,0,0,0.4); /* Black w/ opacity */
+}
+
```
not that you have to, but why not just import the w3css library? it would cut down on the amount of stuff you have to custom make or copy/paste.

mshafer1 reviewed (commented):
Working pretty well.

mshafer1 commented:
I missed that you merged this already, thought it was another push - just take these thoughts into consideration on your next PR.
